File: backend/internal/domain/roadmap.go
package domain

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

const (
	RoadmapVisibilityPrivate = "private"
	RoadmapVisibilityPublic  = "public"
)

type Roadmap struct {
	ID            uuid.UUID `json:"id" gorm:"type:uuid;primaryKey"`
	UserID        uuid.UUID `json:"user_id" gorm:"type:uuid;not null;index"`
	Name          string    `json:"name" gorm:"not null"`
	Description   string    `json:"description" gorm:"type:text"`
	Visibility    string    `json:"visibility" gorm:"type:varchar(10);default:'private'"`
	Topics        []string  `json:"topics" gorm:"type:jsonb;serializer:json;default:'[]'"`
	TotalProblems int       `json:"total_problems" gorm:"default:0"`
	Progress      int       `json:"progress" gorm:"default:0"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`

	User User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

func (r *Roadmap) BeforeCreate(tx *gorm.DB) error {
	r.ID = uuid.New()
	return nil
}

File: frontend/src/types/roadmap/roadmap.ts
export interface ICustomRoadmap {
  id: string;
  user_id: string;
  author_name: string;
  name: string;
  description: string;
  visibility: "private" | "public";
  topics: string[];
  total_problems: number;
  progress: number;
  created_at: string;
  updated_at: string;
}

export interface ICreateCustomRoadmap {
  name: string;
  description: string;
  visibility: "private" | "public";
  topics: string[];
  total_problems: number;
}

export interface IUpdateRoadmapProgress {
  progress: number;
}
